Allow formatting in dask.dataframe.to_hdf path and key parameters

nirizr opened the pull request:
Lets a user save all dask partitions to different files or different hdf datasets in the same hdf file.
It goes well with read_hdf's support for globmatch files and datasets.

Signed-off-by: Nir Izraeli <nirizr@gmail.com>

nirizr commented:
I can see an additional improvement here.
It should be occasionally possible (and improve performance) to remove the links between partition saves, when saving to multiple files for example.

mrocklin commented:
Can you add a test to verify the functionality you'd like to work?  That way you can be assured that future developers won't remove this accidentally.

http://matthewrocklin.com/blog/work/2016/02/08/tests

nirizr commented:
Added a test
removed formatting by name because it has unsupported characters and not really that interesting.
I also think i fixed another test, that was preparing but never actually asserting.

nirizr commented:
Another posiible improvment here could be including the partition names, however i tend to believe this is a more general to_hdf/read_hdf issue.

mrocklin commented:
This looks good to me.  We should probably add this feature to documentation and docstrings, but I can handle that later if you prefer; we'll probably have to create a full docstring for `to_hdf/read_hdf` rather than reusing the pandas one with `wraps`.

Regarding partition names I'm also in favor of sticking to integers as you're doing now.  Alternatively you might like to look at the solution in `dask.bag.to_textfiles` with the `name_function`.  

I plan to merge this soon if there are no further comments.

nirizr commented:
If you say to/read hdfs will be rewritten I won't go there, i've never wrote docstrings so i'll leave it to you.

I like the name_function. I think i'll add, so wait with the merge a bit.

mrocklin commented:
Alright, will do.

If you feel like doing the docstring (again, no pressure) we would probably remove the current decorator around `to_hdf` and copy-and-modify the docstring from pandas, adding and removing keyword arguments as necessary.  Alternatively we could write something fresh.

mrocklin commented:
If we copy and modify we would need to verify that we have a Pandas license on hand.

nirizr commented:
I can't really say I'm comfortable with deciding where/how to place the documentation, but I'm willing to fill it up if someone puts up a skeleton.

Anyway I added a name_function argument with default value of str.format, and I'm good to go.

mrocklin commented on dask/dataframe/tests/test_io.py:
```diff
@@ -697,6 +697,14 @@ def test_to_hdf():
+    with tmpfile('h5') as fn:
+        a.to_hdf(fn, '/data{i}')
+        out = dd.read_hdf(fn, '/*')
+        tm.assert_frame_equal(df, out.compute())
```
Can I ask for a test that uses the name function?

nirizr replied:
Sure thing, coming right up.

nirizr commented:
Adding a test for name_function uncovered an issue. using it effected both file names and dataset names when using file paths (opposed to file handlers).

To handle that, I now use name_function only on dataset names, and only allow str.format() on file names. This seems like an OK tradeoff since I don't think we need to consider complex control over file names a big too useful.

mrocklin commented:
OK, I'm looking through this again more carefully.  Some thoughts:

We should probably share the same interface for `name_function` for both `dask.bag.to_textfiles` and `dask.dataframe.to_hdf`.  If we do this then the following test should probably hold:

```python
    a = dd.from_pandas(df, 2) 
    with tmpfile('h5') as fn:
        a.to_hdf(fn, '/data-*', name_function=lambda i: 'a' * (i +  1))
        out = dd.read_hdf(fn, '/data-*')
        tm.assert_frame_equal(df, out.compute())
                             
        out = pd.read_hdf(fn, '/data-a')
        tm.assert_frame_equal(out, df.iloc[:2])
        out = pd.read_hdf(fn, '/data-aa') 
        tm.assert_frame_equal(out, df.iloc[2:])
```

I like your thought that, if we store the partition to different datapaths then we can be much more parallel.  Currently we force sequential storage by including the  `_link` function into the dask graph.  This makes things more complex than necessary.  It would be nice to switch between two modes of saving files, one that is sequential if `*` is not in the datapath and another that is completely embarrassingly parallel if `*` is in the datapath.

mrocklin commented:
Thoughts?

nirizr commented:
Ill give the test a look when I'm home, should be fine.

Regarding the optimization, yeah, that was my main goal here. I mentioned that in my second comment here. We can let the user decide on the modes, but it'll be way cooler achieving that conclusion outomatically, taking into account formatting could collide.

I thought about getting there a bit later, because I prefer having this in for the next release before aiming further, so I suggest we seperate that to different PRs. Later on this might get to other formats,

I also contemplated doing a two phase saves, first to multiple.temporary hdfs, merging them into one later.

mrocklin commented:
Moving parallel writes to a separate PR sounds good to me.  I do think that we should make sure that the interfaces between `to_textfiles` and `to_hdf` should match in this PR.  This would require dumbing down the `name_function` in this PR a bit.  `db.to_textfiles` uses `name_function=str` while this PR uses `name_function=str.format` with conventions using `key`, `name`, and `i`.  I'm tempted by the simplicity of using `name_function=str` for now and setting `datapath = datapath.replace('*', name_function(i))` for each key.

nirizr commented:
Sorry, I wasn't clear enough. I'll be simplifying the PR as you suggested in a couple of hours. I wasn't too sure about it myself and you're right about keeping it simple.

nirizr commented:
I simplified the name_function as requested and added the suggested test.
I also added some additional tests (related to saving to multiple h5 files), and added a tmpdir contextmanager to dask.utils for these tests.

mrocklin commented on dask/dataframe/tests/test_io.py:
```diff
@@ -697,6 +697,61 @@ def test_to_hdf():
+        out = pd.read_hdf(os.path.join(dn, 'data_a.h5'), '/data_a')
+        tm.assert_frame_equal(out, df.iloc[:2])
+        out = pd.read_hdf(os.path.join(dn, 'data_aa.h5'), '/data_aa')
+        tm.assert_frame_equal(out, df.iloc[2:])
```
This last test feels a bit odd.  It's not clear to me what proper behavior is when both the path and the datapath have a `*` in them.  I think it's unlikely that the user wants the same naming scheme used in both.  I would probably opt to err in this case.

nirizr replied:
It started by me thinking what will happen in such a case, when i saw the test passed i just kept it, I wanted to make sure it's addressed.
I agree there's no clear use case for that, but someone might enjoy it in the future.
I'm removing it, let me know if you change your mind.

mrocklin commented:
Other than my recent comment this looks good to me.  The py27 exception is unrelated to this work and should be cleaned up shortly.

nirizr commented:
I think we're good to go, please let me know if there's anything else you need

mrocklin commented:
Agreed.   This looks great to me.  Merging.

nirizr commented:
\o/
